File: Python/ADMM_CNJGL/ADM_CNJGL.py
# ADMM algorithm for CNJGL

import ADM_parameters
from numpy.linalg import norm
from expand_ni import expand_ni
from soft_thresh import soft_thresh
from soft_scal import soft_scal
from diag_construction import diag_construction
from nodiag_construction import nodiag_construction

from numpy import eye, zeros, diag, matrix, concatenate

def ADM_CNJGL(S1, S2, lam1, lam2, n1, n2):
    # Begin algorithm
    p = S1.shape[0]
    
    # Initialize variables
    Theta1old = eye(p)
    Theta2old = eye(p)
    Theta1 = eye(p)
    Theta2 = eye(p)
    V1 = Theta1old/2
    V2 = Theta2old/2
    W1 = V1.transpose()
    W2 = V2.transpose()
    Z1 = Theta1
    Z2 = Theta2
    Gamma1 = matrix(zeros((p,p)))
    Gamma2 = matrix(zeros((p,p)))
    Lambda1 = matrix(zeros((p,p)))
    Lambda2 = matrix(zeros((p,p)))
    Q1 = matrix(zeros((p,p)))
    Q2 = Q1

    # Load parameters
    rho = ADM_parameters.opts['rho_init']
    iter_adm = matrix(zeros((ADM_parameters.opts['homotopy_size'], 1)))
    
    relError = matrix(zeros((ADM_parameters.opts['maxiter'] + ADM_parameters.opts['homotopy_size']*ADM_parameters.opts['maxiter'], 1)))

    print "homotopy_size", int(ADM_parameters.opts['homotopy_size'])
    print "maxiter", int(ADM_parameters.opts['maxiter'])

    # Begin outer loop for homotopy parameter
    for r in range(0, int(ADM_parameters.opts['homotopy_size'])): # 5
        # Begin ADM
        for i in range(0, ADM_parameters.opts['maxiter']): # 500
            # Update of Theta1 and Theta2

            Theta1 = expand_ni(1.0/(2.0*rho)*(rho*(V1 + W1 + Z1) - (Gamma1 + Lambda1)), S1, rho, n1)
            Theta2 = expand_ni(1.0/(2.0*rho)*(rho*(V2 + W2 + Z2) - (Gamma2 + Lambda2)), S2, rho, n2)
            
            # print "Theta1\n", Theta1

            # Update of Z1 and Z2
            Z1 = soft_thresh(Theta1 + Lambda1/(rho+0.0), lam1/(rho+0.0))
            Z2 = soft_thresh(Theta2 + Lambda2/(rho+0.0), lam1/(rho+0.0))

            # Update of V1 and V2
            C1 = 1/(2.0*rho)*(rho*(W1.transpose() + Theta1 - W1) - (Q1 - Gamma1))
            C2 = 1/(2.0*rho)*(rho*(W2.transpose() + Theta2 - W2) - (Q2 - Gamma2));

#            print "C1", C1.shape
            d1 = diag(C1)
            d2 = diag(C2)
            N1 = nodiag_construction(C1)
            N2 = nodiag_construction(C2)
#            print "N1", N1.shape
            N = concatenate((N1,N2))
#            print "N", N.shape
            H = soft_scal(N,lam2/(2.0*rho))

            d = concatenate((d1,d2))
#            print "d1", d1.shape
#            print "H", H.shape
#            print "diagcon H", H[0:p-1,:].shape
            V1 = diag_construction(H[0:p-1,:],d1);
#            print "diagcon2 H", H[p-1:,:].shape
            V2 = diag_construction(H[p-1:,:],d2); # err?

            # Update of W1 and W2
            W1 = 1/(2.0*rho)*(rho*(V1.transpose() + Theta1 - V1) + Gamma1 + Q1.transpose());
            W2 = 1/(2.0*rho)*(rho*(V2.transpose() + Theta2 - V2) + Gamma2 + Q2.transpose());

            # Update of dual variables
            Gamma1 = Gamma1 + rho*(Theta1 - V1 - W1);
            Gamma2 = Gamma2 + rho*(Theta2 - V2 - W2);
            Lambda1 = Lambda1 + rho*(Theta1 - Z1);
            Lambda2 = Lambda2 + rho*(Theta2 - Z2);
            Q1 = Q1 + rho*(V1 - W1.transpose());
            Q2 = Q2 + rho*(V2 - W2.transpose());

            norm1 =  norm(Theta1 - Theta1old, 'fro') / norm(Theta1, 'fro')
            norm2 =  norm(Theta2 - Theta2old, 'fro') / norm(Theta2, 'fro')

            relError[i + r*ADM_parameters.opts['maxiter'], 0] = \
                max(norm(Theta1 - Theta1old, 'fro') / norm(Theta1, 'fro'), \
                    norm(Theta2 - Theta2old, 'fro') / norm(Theta2, 'fro'))
            
            # print i, r, relError[i + r*ADM_parameters.opts['maxiter'], 0]

            if relError[i + (r*ADM_parameters.opts['maxiter']), 0] <= ADM_parameters.opts['eps']:
                break
            
            Theta1old = Theta1
            Theta2old = Theta2
            # End ADM

    rho = rho * ADM_parameters.opts['homotopy']
    iter_adm[r,0] = i
    return (Theta1, Theta2, V1, V2, iter_adm, relError)
# End algorithm

File: CNJGL_onweb/ADMM_CNJGL/ADM_CNJGL.m
%%--------------- ADMM algorithm for CNJGL -------------- %%
% ------------------- KARTHIK MOHAN, UW ------------------ %

% ------------------------ LAST UPDATE: 8/1/2013 ---------------------- %



function [Theta_1,Theta_2,V_1,V_2,iter_adm, relError] = ADM_CNJGL(S_1,S_2,lambda_1,lambda_2,n_1,n_2)


%------------------------------
% BEGIN ALGORITHM
%------------------------------
p = size(S_1,1);

%----------------------
% INITIALIZE VARIABLES
%----------------------
Theta_1_old = eye(p); Theta_2_old = eye(p);
Theta_1 = eye(p); Theta_2 = eye(p);
V_1 = Theta_1_old/2; V_2 = Theta_2_old/2;
W_1 = V_1'; W_2 = V_2'; 
Z_1 = Theta_1; Z_2 = Theta_2;
Gamma_1 = zeros(p,p); Gamma_2 = Gamma_1; 
Lambda_1 = zeros(p,p); Lambda_2 = zeros(p,p); 
Q_1 = zeros(p,p); Q_2 = Q_1; 
%}
%----------------------
% LOAD PARAMETERS 
%----------------------
ADM_parameters;
rho = opts.rho_init;
iter_adm = zeros(opts.homotopy_size,1);

relError = zeros(opts.maxiter, 1);

%----------------------------------------
% Begin outer loop for homotopy parameter
%----------------------------------------

disp ("homotopy_size")
disp (opts.homotopy_size)

disp ("maxiter")
disp (opts.maxiter)

for (r = 1:opts.homotopy_size)
    
    %---------------------
    % Begin ADM 
    %---------------------
 for (i = 1:opts.maxiter)
     %Update 8 primal variables and 6 dual variables in the ADM formulation

     % Update of Theta_1 and Theta_2
     Theta_1 = expand_ni(1/(2*rho)*(rho*(V_1 + W_1 + Z_1) - (Gamma_1 + Lambda_1)),S_1,rho,n_1);
     Theta_2 = expand_ni(1/(2*rho)*(rho*(V_2 + W_2 + Z_2) - (Gamma_2 + Lambda_2)),S_2,rho,n_2);

     % Update of Z_1 and Z_2
     Z_1 = soft_thresh(Theta_1 + Lambda_1/rho, lambda_1/rho);
     Z_2 = soft_thresh(Theta_2 + Lambda_2/rho, lambda_1/rho);

     % UPDATE OF V_1 and V_2
     C_1 = 1/(2*rho)*(rho*(W_1' + Theta_1 - W_1) - (Q_1 - Gamma_1));
     C_2 = 1/(2*rho)*(rho*(W_2' + Theta_2 - W_2) - (Q_2 - Gamma_2));

%     disp("C_1")
%     disp(size(C_1))
     d_1 = diag(C_1); d_2 = diag(C_2);
     N_1 = nodiag_construction(C_1);
     N_2 = nodiag_construction(C_2);
%     disp("N_1")
%     disp(size(N_1))
     N = [N_1; N_2];
%     disp("N")
%     disp(size(N))
     H = soft_scal(N,lambda_2/(2*rho)); d = [d_1;d_2];

%     disp("d_1")
%     disp(size(d_1))
%     disp("H")
%     disp(size(H))

%     disp("diagcon H")
%     disp(size(H(1:p-1,:)))

     V_1 = diag_construction(H(1:p-1,:),d_1);

%     disp("diagcon2 H")
%     disp(size(H(p:2*p-2,:)))

     V_2 = diag_construction(H(p:2*p-2,:),d_2);

     % Update of W_1 and W_2
     W_1 = 1/(2*rho)*(rho*(V_1' + Theta_1 - V_1) + Gamma_1 + Q_1');
     W_2 = 1/(2*rho)*(rho*(V_2' + Theta_2 - V_2) + Gamma_2 + Q_2');

     % Update of Dual variables
     Gamma_1 = Gamma_1 + rho*(Theta_1 - V_1 - W_1);
     Gamma_2 = Gamma_2 + rho*(Theta_2 - V_2 - W_2);
     Lambda_1 = Lambda_1 + rho*(Theta_1 - Z_1);
     Lambda_2 = Lambda_2 + rho*(Theta_2 - Z_2);
     Q_1 = Q_1 + rho*(V_1 - W_1');
     Q_2 = Q_2 + rho*(V_2 - W_2');

%     disp("i")
%     disp(i)
%     disp("r")
%     disp(r)
%     disp("relError index")
%     disp(i + ((r - 1)*opts.maxiter))

     relError(i + ((r - 1)*opts.maxiter), 1) = ...
         max(norm(Theta_1 - Theta_1_old,'fro')/norm(Theta_1,'fro'), ...
             norm(Theta_2 - Theta_2_old,'fro')/norm(Theta_2,'fro'));

    if(relError(i + ((r - 1)*opts.maxiter), 1) <= opts.eps)
        break;
    end;
    Theta_1_old = Theta_1; Theta_2_old = Theta_2;
 end;
     %---------------------
     % END ADM 
     %---------------------
 
rho = rho*opts.homotopy;
iter_adm(r,1) = i;
end;


%-----------------------------
% END ALGORITHM
%-----------------------------
